diff --git a/drivers/media/i2c/max9295.c b/drivers/media/i2c/max9295.c
index 12b5db00..5f9fd782 100644
--- a/drivers/media/i2c/max9295.c
+++ b/drivers/media/i2c/max9295.c
@@ -109,6 +109,8 @@ struct max9295 {
 	struct i2c_client *i2c_client;
 	struct regmap *regmap;
 	struct max9295_client_ctx g_client;
+	struct v4l2_subdev sd;
+	struct media_pad pad[MAX9295_MAX_PIPES + 1];
 	struct mutex lock;
 	bool external_addr_reassign;
 	/* primary serializer properties */
@@ -141,6 +143,440 @@ static int max9295_write_reg(struct device *dev, u16 addr, u8 val)
 	return err;
 }
 
+struct v4l2_subdev_ops max_ops;
+
+static int max_power_on(struct v4l2_subdev *sd, int on)
+{
+	struct max9295 *priv = container_of(sd, struct max9295, sd);
+	dev_info(&priv->i2c_client->dev, "%s called\n", __func__);
+    return 0;
+}
+
+static int max_reset_control(struct v4l2_subdev *sd, u32 val)
+{
+	/* TODO EHUD: Is needed? */
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_log_status(struct v4l2_subdev *sd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_io_pin_config(struct v4l2_subdev *sd, size_t n,
+		struct v4l2_subdev_io_pin_config *pincfg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_init(struct v4l2_subdev *sd, u32 val) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_load_fw(struct v4l2_subdev *sd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_reset(struct v4l2_subdev *sd, u32 val) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_gpio(struct v4l2_subdev *sd, u32 val) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static long max_command(struct v4l2_subdev *sd, unsigned int cmd, void *arg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static long max_ioctl(struct v4l2_subdev *sd, unsigned int cmd, void *arg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+#ifdef CONFIG_COMPAT
+static long max_compat_ioctl32(struct v4l2_subdev *sd, unsigned int cmd,
+	unsigned long arg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+	}
+#endif
+#ifdef CONFIG_VIDEO_ADV_DEBUG
+static int max_g_register)(struct v4l2_subdev *sd, struct v4l2_dbg_register *reg) {
+	dev_info(sd->dev, "%s called\n", __func__) {
+	return 0;
+	}
+static int max_s_register)(struct v4l2_subdev *sd, const struct v4l2_dbg_register *reg) {
+	dev_info(sd->dev, "%s called\n", __func__) {
+	return 0;
+	}
+#endif
+static int max_s_power(struct v4l2_subdev *sd, int on) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+	}
+
+static int max_interrupt_service_routine(struct v4l2_subdev *sd,
+	u32 status, bool *handled) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_subscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
+	struct v4l2_event_subscription *sub) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_unsubscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
+		struct v4l2_event_subscription *sub) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_sync(struct v4l2_subdev *sd, unsigned int sync_events) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+struct v4l2_subdev_core_ops max_core_ops = {
+	.log_status = max_log_status,
+	.s_io_pin_config = max_s_io_pin_config,
+	.init = max_init,
+	.load_fw = max_load_fw,
+	.reset = max_reset_control,
+	.s_gpio = max_s_gpio,
+	.command = max_command,
+	.ioctl = max_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl32 = max_compat_ioctl32,
+#endif
+#ifdef CONFIG_VIDEO_ADV_DEBUG
+	.g_register = max_g_register,
+	.s_register = max_s_register,
+#endif
+	.s_power = max_power_on,
+	.interrupt_service_routine = max_interrupt_service_routine,
+	.subscribe_event = max_subscribe_event,
+	.unsubscribe_event = max_unsubscribe_event,
+	.sync = max_sync
+};
+
+static int max_s_routing(struct v4l2_subdev *sd, u32 input, u32 output, u32 config) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_crystal_freq(struct v4l2_subdev *sd, u32 freq, u32 flags) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_std(struct v4l2_subdev *sd, v4l2_std_id *norm) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_std(struct v4l2_subdev *sd, v4l2_std_id norm) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_std_output(struct v4l2_subdev *sd, v4l2_std_id std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_std_output(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_querystd(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_tvnorms(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_tvnorms_output(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_input_status(struct v4l2_subdev *sd, u32 *status) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_stream(struct v4l2_subdev *sd, int enable) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_pixelaspect(struct v4l2_subdev *sd, struct v4l2_fract *aspect) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_dv_timings(struct v4l2_subdev *sd,
+			struct v4l2_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_dv_timings(struct v4l2_subdev *sd,
+			struct v4l2_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_query_dv_timings(struct v4l2_subdev *sd,
+			struct v4l2_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_rx_buffer(struct v4l2_subdev *sd, void *buf,
+			   unsigned int *size) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_pre_streamon(struct v4l2_subdev *sd, u32 flags) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_post_streamoff(struct v4l2_subdev *sd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+struct v4l2_subdev_video_ops max_video_ops = {
+	.s_routing = max_s_routing,
+	.s_crystal_freq = max_s_crystal_freq,
+	.g_std = max_g_std,
+	.s_std = max_s_std,
+	.s_std_output = max_s_std_output,
+	.g_std_output = max_g_std_output,
+	.querystd = max_querystd,
+	.g_tvnorms = max_g_tvnorms,
+	.g_tvnorms_output = max_g_tvnorms_output,
+	.g_input_status = max_g_input_status,
+	.s_stream = max_s_stream,
+	.g_pixelaspect = max_g_pixelaspect,
+	.s_dv_timings = max_s_dv_timings,
+	.g_dv_timings = max_g_dv_timings,
+	.query_dv_timings = max_query_dv_timings,
+	.s_rx_buffer = max_s_rx_buffer,
+	.pre_streamon = max_pre_streamon,
+	.post_streamoff = max_post_streamoff
+};
+
+static int max_enum_mbus_code(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_mbus_code_enum *code) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enum_frame_size(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_frame_size_enum *fse) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enum_frame_interval(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_frame_interval_enum *fie) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_fmt(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_format *format) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_fmt(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_format *format) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_selection(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_selection *sel) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_selection(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_selection *sel) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_frame_interval(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_frame_interval *interval) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+	static int max_set_frame_interval(struct v4l2_subdev *sd,
+			struct v4l2_subdev_state *state,
+			struct v4l2_subdev_frame_interval *interval) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_edid(struct v4l2_subdev *sd, struct v4l2_edid *edid) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_edid(struct v4l2_subdev *sd, struct v4l2_edid *edid) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_dv_timings_cap(struct v4l2_subdev *sd,
+		struct v4l2_dv_timings_cap *cap) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enum_dv_timings(struct v4l2_subdev *sd,
+		struct v4l2_enum_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+#ifdef CONFIG_MEDIA_CONTROLLER
+static int max_link_validate(struct v4l2_subdev *sd, struct media_link *link,
+		struct v4l2_subdev_format *source_fmt,
+		struct v4l2_subdev_format *sink_fmt) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+#endif /* CONFIG_MEDIA_CONTROLLER */
+static int max_get_frame_desc(struct v4l2_subdev *sd, unsigned int pad,
+		struct v4l2_mbus_frame_desc *fd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_frame_desc(struct v4l2_subdev *sd, unsigned int pad,
+		struct v4l2_mbus_frame_desc *fd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_mbus_config(struct v4l2_subdev *sd, unsigned int pad,
+		struct v4l2_mbus_config *config) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_routing(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		enum v4l2_subdev_format_whence which,
+		struct v4l2_subdev_krouting *route) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enable_streams(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state, u32 pad,
+		u64 streams_mask) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_disable_streams(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state, u32 pad,
+		u64 streams_mask) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+struct v4l2_subdev_pad_ops max_pad_ops = {
+	.enum_mbus_code = max_enum_mbus_code,
+	.enum_frame_size = max_enum_frame_size,
+	.enum_frame_interval = max_enum_frame_interval,
+	.get_fmt = max_get_fmt,
+	.set_fmt = max_set_fmt,
+	.get_selection = max_get_selection,
+	.set_selection = max_set_selection,
+	.get_frame_interval = max_get_frame_interval,
+	.set_frame_interval = max_set_frame_interval,
+	.get_edid = max_get_edid,
+	.set_edid = max_set_edid,
+	.dv_timings_cap = max_dv_timings_cap,
+	.enum_dv_timings = max_enum_dv_timings,
+#ifdef CONFIG_MEDIA_CONTROLLER
+	.link_validate = max_link_validate,
+#endif /* CONFIG_MEDIA_CONTROLLER */
+	.get_frame_desc = max_get_frame_desc,
+	.set_frame_desc = max_set_frame_desc,
+	.get_mbus_config = max_get_mbus_config,
+	.set_routing = max_set_routing,
+	.enable_streams = max_enable_streams,
+	.disable_streams = max_disable_streams
+};
+struct v4l2_subdev_ops max_ops = {
+	.core = &max_core_ops,
+	.tuner = NULL,
+	.audio = NULL,
+	.video = &max_video_ops,
+	.vbi = NULL,
+	.ir = NULL,
+	.sensor = NULL,
+	.pad = &max_pad_ops
+};
+
+static int max_sdev_register(struct max9295 *priv)
+{
+	struct media_entity *entity = &priv->sd.entity;
+	struct media_pad *pad = priv->pad;
+	int ret;
+	dev_info(&priv->i2c_client->dev, "%s called\n", __func__);
+	priv->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
+	for (int i = 1; i < MAX9295_MAX_PIPES + 1; i++)
+		pad[i].index = MEDIA_PAD_FL_SINK;
+	pad->flags = MEDIA_PAD_FL_SOURCE;
+
+	media_entity_pads_init(entity, MAX9295_MAX_PIPES + 1, pad);
+	entity->obj_type = MEDIA_ENTITY_TYPE_V4L2_SUBDEV;
+	entity->function = MEDIA_ENT_F_VID_MUX;
+
+	v4l2_subdev_init(&priv->sd, &max_ops);
+	priv->sd.owner = THIS_MODULE;
+	ret = v4l2_async_register_subdev(&priv->sd);
+	return ret;
+}
+
+static int max_sdev_unregister(struct max9295 *priv)
+{
+	v4l2_async_register_subdev(&priv->sd);
+	dev_info(&priv->i2c_client->dev, "%s called\n", __func__);
+	return 0;
+}
+
 int max9295_setup_streaming(struct device *dev)
 {
 	struct max9295 *priv = dev_get_drvdata(dev);
@@ -274,7 +710,6 @@ error:
 	mutex_unlock(&priv->lock);
 	return err;
 }
-EXPORT_SYMBOL(max9295_setup_streaming);
 
 int max9295_setup_control(struct device *dev)
 {
@@ -379,7 +814,6 @@ error:
 	mutex_unlock(&priv->lock);
 	return err;
 }
-EXPORT_SYMBOL(max9295_setup_control);
 
 int max9295_setup_gpio_tunneling(struct device *dev)
 {
@@ -388,7 +822,7 @@ int max9295_setup_gpio_tunneling(struct device *dev)
 
 	mutex_lock(&priv->lock);
 
-	/* Configure ESYNC registers for MAX96712A deserializer */
+	/* Configure ESYNC registers for MAX9295 deserializer */
 	max9295_write_reg(dev, MAX9295_SRC_PWDN_ADDR, MAX9295_PWDN_ESYNC_EXT);
 	max9295_write_reg(dev, MAX9295_SRC_CTRL_ADDR, MAX9295_RESET_ESYNC);
 	max9295_write_reg(dev, MAX9295_2C0_ADDR, MAX9295_2C0_ESYNC);
@@ -400,7 +834,6 @@ int max9295_setup_gpio_tunneling(struct device *dev)
 
 	return err;
 }
-EXPORT_SYMBOL(max9295_setup_gpio_tunneling);
 
 int max9295_reset_control(struct device *dev)
 {
@@ -430,7 +863,6 @@ error:
 	mutex_unlock(&priv->lock);
 	return err;
 }
-EXPORT_SYMBOL(max9295_reset_control);
 
 int max9295_sdev_pair(struct device *dev, struct gmsl_link_ctx *g_ctx)
 {
@@ -458,7 +890,6 @@ error:
 	mutex_unlock(&priv->lock);
 	return 0;
 }
-EXPORT_SYMBOL(max9295_sdev_pair);
 
 int max9295_sdev_unpair(struct device *dev, struct device *s_dev)
 {
@@ -493,7 +924,6 @@ error:
 	mutex_unlock(&priv->lock);
 	return err;
 }
-EXPORT_SYMBOL(max9295_sdev_unpair);
 
 static  struct regmap_config max9295_regmap_config = {
 	.reg_bits = 16,
@@ -605,7 +1035,6 @@ int max9295_init_settings(struct device *dev)
 
 	return err;
 }
-EXPORT_SYMBOL(max9295_init_settings);
 
 int max9295_set_pipe(struct device *dev, int pipe_id,
 		     u8 data_type1, u8 data_type2, u32 vc_id)
@@ -630,7 +1059,6 @@ int max9295_set_pipe(struct device *dev, int pipe_id,
 
 	return err;
 }
-EXPORT_SYMBOL(max9295_set_pipe);
 
 #if defined(NV_I2C_DRIVER_STRUCT_PROBE_WITHOUT_I2C_DEVICE_ID_ARG) /* Linux 6.3 */
 static int max9295_probe(struct i2c_client *client)
@@ -681,6 +1109,8 @@ static int max9295_probe(struct i2c_client *client,
 
 	dev_set_drvdata(&client->dev, priv);
 
+	max_sdev_register(priv);
+
 	/* dev communication gets validated when GMSL link setup is done */
 	dev_info(&client->dev, "%s: success\n", __func__);
 
@@ -694,9 +1124,9 @@ static void max9295_remove(struct i2c_client *client)
 #endif
 {
 	struct max9295 *priv;
-
 	if (client != NULL) {
 		priv = dev_get_drvdata(&client->dev);
+		max_sdev_unregister(priv);
 		mutex_destroy(&priv->lock);
 		i2c_unregister_device(client);
 		client = NULL;
diff --git a/drivers/media/i2c/max96712.c b/drivers/media/i2c/max96712.c
index e37fe512..79026397 100644
--- a/drivers/media/i2c/max96712.c
+++ b/drivers/media/i2c/max96712.c
@@ -349,73 +349,384 @@ static int max_reset_control(struct v4l2_subdev *sd, u32 val)
 	return 0;
 }
 
+static int max_log_status(struct v4l2_subdev *sd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_io_pin_config(struct v4l2_subdev *sd, size_t n,
+		struct v4l2_subdev_io_pin_config *pincfg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_init(struct v4l2_subdev *sd, u32 val) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_load_fw(struct v4l2_subdev *sd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_reset(struct v4l2_subdev *sd, u32 val) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_gpio(struct v4l2_subdev *sd, u32 val) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static long max_command(struct v4l2_subdev *sd, unsigned int cmd, void *arg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static long max_ioctl(struct v4l2_subdev *sd, unsigned int cmd, void *arg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+#ifdef CONFIG_COMPAT
+static long max_compat_ioctl32(struct v4l2_subdev *sd, unsigned int cmd,
+	unsigned long arg) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+	}
+#endif
+#ifdef CONFIG_VIDEO_ADV_DEBUG
+static int max_g_register)(struct v4l2_subdev *sd, struct v4l2_dbg_register *reg) {
+	dev_info(sd->dev, "%s called\n", __func__) {
+	return 0;
+	}
+static int max_s_register)(struct v4l2_subdev *sd, const struct v4l2_dbg_register *reg) {
+	dev_info(sd->dev, "%s called\n", __func__) {
+	return 0;
+	}
+#endif
+static int max_s_power(struct v4l2_subdev *sd, int on) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+	}
+
+static int max_interrupt_service_routine(struct v4l2_subdev *sd,
+	u32 status, bool *handled) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_subscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
+	struct v4l2_event_subscription *sub) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_unsubscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
+		struct v4l2_event_subscription *sub) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_sync(struct v4l2_subdev *sd, unsigned int sync_events) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
 struct v4l2_subdev_core_ops max_core_ops = {
-	.log_status = NULL,
-	.s_io_pin_config = NULL,
-	.init = NULL,
-	.load_fw = NULL,
+	.log_status = max_log_status,
+	.s_io_pin_config = max_s_io_pin_config,
+	.init = max_init,
+	.load_fw = max_load_fw,
 	.reset = max_reset_control,
-	.s_gpio = NULL,
-	.command = NULL,
-	.ioctl = NULL,
+	.s_gpio = max_s_gpio,
+	.command = max_command,
+	.ioctl = max_ioctl,
 #ifdef CONFIG_COMPAT
-	.compat_ioctl32 = NULL,
+	.compat_ioctl32 = max_compat_ioctl32,
 #endif
 #ifdef CONFIG_VIDEO_ADV_DEBUG
-	.g_register = NULL,
-	.s_register = NULL,
+	.g_register = max_g_register,
+	.s_register = max_s_register,
 #endif
 	.s_power = max_power_on,
-	.interrupt_service_routine = NULL,
-	.subscribe_event = NULL,
-	.unsubscribe_event = NULL,
-	.sync = NULL
+	.interrupt_service_routine = max_interrupt_service_routine,
+	.subscribe_event = max_subscribe_event,
+	.unsubscribe_event = max_unsubscribe_event,
+	.sync = max_sync
 };
 
+static int max_s_routing(struct v4l2_subdev *sd, u32 input, u32 output, u32 config) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_crystal_freq(struct v4l2_subdev *sd, u32 freq, u32 flags) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_std(struct v4l2_subdev *sd, v4l2_std_id *norm) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_std(struct v4l2_subdev *sd, v4l2_std_id norm) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_std_output(struct v4l2_subdev *sd, v4l2_std_id std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_std_output(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_querystd(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_tvnorms(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_tvnorms_output(struct v4l2_subdev *sd, v4l2_std_id *std) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_input_status(struct v4l2_subdev *sd, u32 *status) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_stream(struct v4l2_subdev *sd, int enable) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_pixelaspect(struct v4l2_subdev *sd, struct v4l2_fract *aspect) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_dv_timings(struct v4l2_subdev *sd,
+			struct v4l2_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_g_dv_timings(struct v4l2_subdev *sd,
+			struct v4l2_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_query_dv_timings(struct v4l2_subdev *sd,
+			struct v4l2_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_s_rx_buffer(struct v4l2_subdev *sd, void *buf,
+			   unsigned int *size) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_pre_streamon(struct v4l2_subdev *sd, u32 flags) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_post_streamoff(struct v4l2_subdev *sd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
 struct v4l2_subdev_video_ops max_video_ops = {
-	.s_routing = NULL,
-	.s_crystal_freq = NULL,
-	.g_std = NULL,
-	.s_std = NULL,
-	.s_std_output = NULL,
-	.g_std_output = NULL,
-	.querystd = NULL,
-	.g_tvnorms = NULL,
-	.g_tvnorms_output = NULL,
-	.g_input_status = NULL,
-	.s_stream = NULL,
-	.g_pixelaspect = NULL,
-	.s_dv_timings = NULL,
-	.g_dv_timings = NULL,
-	.query_dv_timings = NULL,
-	.s_rx_buffer = NULL,
-	.pre_streamon = NULL,
-	.post_streamoff = NULL
+	.s_routing = max_s_routing,
+	.s_crystal_freq = max_s_crystal_freq,
+	.g_std = max_g_std,
+	.s_std = max_s_std,
+	.s_std_output = max_s_std_output,
+	.g_std_output = max_g_std_output,
+	.querystd = max_querystd,
+	.g_tvnorms = max_g_tvnorms,
+	.g_tvnorms_output = max_g_tvnorms_output,
+	.g_input_status = max_g_input_status,
+	.s_stream = max_s_stream,
+	.g_pixelaspect = max_g_pixelaspect,
+	.s_dv_timings = max_s_dv_timings,
+	.g_dv_timings = max_g_dv_timings,
+	.query_dv_timings = max_query_dv_timings,
+	.s_rx_buffer = max_s_rx_buffer,
+	.pre_streamon = max_pre_streamon,
+	.post_streamoff = max_post_streamoff
 };
 
+static int max_enum_mbus_code(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_mbus_code_enum *code) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enum_frame_size(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_frame_size_enum *fse) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enum_frame_interval(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_frame_interval_enum *fie) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_fmt(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_format *format) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_fmt(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_format *format) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_selection(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_selection *sel) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_selection(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_selection *sel) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_frame_interval(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		struct v4l2_subdev_frame_interval *interval) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+	static int max_set_frame_interval(struct v4l2_subdev *sd,
+			struct v4l2_subdev_state *state,
+			struct v4l2_subdev_frame_interval *interval) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_edid(struct v4l2_subdev *sd, struct v4l2_edid *edid) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_edid(struct v4l2_subdev *sd, struct v4l2_edid *edid) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_dv_timings_cap(struct v4l2_subdev *sd,
+		struct v4l2_dv_timings_cap *cap) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enum_dv_timings(struct v4l2_subdev *sd,
+		struct v4l2_enum_dv_timings *timings) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+#ifdef CONFIG_MEDIA_CONTROLLER
+static int max_link_validate(struct v4l2_subdev *sd, struct media_link *link,
+		struct v4l2_subdev_format *source_fmt,
+		struct v4l2_subdev_format *sink_fmt) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+#endif /* CONFIG_MEDIA_CONTROLLER */
+static int max_get_frame_desc(struct v4l2_subdev *sd, unsigned int pad,
+		struct v4l2_mbus_frame_desc *fd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_frame_desc(struct v4l2_subdev *sd, unsigned int pad,
+		struct v4l2_mbus_frame_desc *fd) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_get_mbus_config(struct v4l2_subdev *sd, unsigned int pad,
+		struct v4l2_mbus_config *config) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_set_routing(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state,
+		enum v4l2_subdev_format_whence which,
+		struct v4l2_subdev_krouting *route) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_enable_streams(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state, u32 pad,
+		u64 streams_mask) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
+static int max_disable_streams(struct v4l2_subdev *sd,
+		struct v4l2_subdev_state *state, u32 pad,
+		u64 streams_mask) {
+	dev_info(sd->dev, "%s called\n", __func__);
+	return 0;
+}
+
 struct v4l2_subdev_pad_ops max_pad_ops = {
-	.enum_mbus_code = NULL,
-	.enum_frame_size = NULL,
-	.enum_frame_interval = NULL,
-	.get_fmt = NULL,
-	.set_fmt = NULL,
-	.get_selection = NULL,
-	.set_selection = NULL,
-	.get_frame_interval = NULL,
-	.set_frame_interval = NULL,
-	.get_edid = NULL,
-	.set_edid = NULL,
-	.dv_timings_cap = NULL,
-	.enum_dv_timings = NULL,
+	.enum_mbus_code = max_enum_mbus_code,
+	.enum_frame_size = max_enum_frame_size,
+	.enum_frame_interval = max_enum_frame_interval,
+	.get_fmt = max_get_fmt,
+	.set_fmt = max_set_fmt,
+	.get_selection = max_get_selection,
+	.set_selection = max_set_selection,
+	.get_frame_interval = max_get_frame_interval,
+	.set_frame_interval = max_set_frame_interval,
+	.get_edid = max_get_edid,
+	.set_edid = max_set_edid,
+	.dv_timings_cap = max_dv_timings_cap,
+	.enum_dv_timings = max_enum_dv_timings,
 #ifdef CONFIG_MEDIA_CONTROLLER
-	.link_validate = NULL,
+	.link_validate = max_link_validate,
 #endif /* CONFIG_MEDIA_CONTROLLER */
-	.get_frame_desc = NULL,
-	.set_frame_desc = NULL,
-	.get_mbus_config = NULL,
-	.set_routing = NULL,
-	.enable_streams = NULL,
-	.disable_streams = NULL
+	.get_frame_desc = max_get_frame_desc,
+	.set_frame_desc = max_set_frame_desc,
+	.get_mbus_config = max_get_mbus_config,
+	.set_routing = max_set_routing,
+	.enable_streams = max_enable_streams,
+	.disable_streams = max_disable_streams
 };
 struct v4l2_subdev_ops max_ops = {
 	.core = &max_core_ops,
